csharp/ql/src/semmle/code/csharp/Unification.qll

```diff
@@ -10,9 +10,69 @@ private import Caching
  * equal modulo identity conversions and type parameters.
  */
 module Gvn {
+  /** Gets the qualified name of type `t`. */
+  string getQualifiedName(Type t) {
+    if not t instanceof NestedType or t.(NestedType).getDeclaringType() instanceof GenericType
+    then result = t.getName()
+    else result = getQualifiedName(t.(NestedType).getDeclaringType()) + "." + t.getName()
+  }
+
+  /**
+   * A generic type. This is either a type with a type parameter, a type with
+   * a type argument, or a nested type with a generic enclosing type.
+   */
+  class GenericType extends Type {
+    GenericType() {
+      exists(this.getChild(0))
+      or
+      this.(NestedType).getDeclaringType() instanceof GenericType
+    }
+
+    /** Gets the generic containing type, if any. */
+    GenericType getQualifier() { result = this.(NestedType).getDeclaringType() }
```

Comment:
What's a little awkward about this is that this only yields generics, rather than the fully qualified type name. I understand that you don't really care about non-generic containing types, but the interface seems a little odd.

Reply:
What if I change the name to `getGenericDeclaringType`?